### PR TITLE
SNOW-1915469 Basic support for DECFLOAT type

### PR DESCRIPTION
**Pull request by sfc-gh-tkissinger**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Implements [SNOW-1915469](https://snowflakecomputing.atlassian.net/browse/SNOW-1915469)

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [x] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR adds basic decimal float (DECFLOAT) data type support to the python connector.
- Supports ARROW and JSON result format. DECFLOATs are converted to Decimal or numpy.float64
- Server-side binding support using explicit Snowflake type specification as Decimal is already mapped to FIXED by default
- No Pandas support

4. (Optional) PR for stored-proc connector:

[SNOW-1915469]: https://snowflakecomputing.atlassian.net/browse/SNOW-1915469?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sfc-gh-tkissinger** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/snowflake/connector/arrow_context.py [hunk 1]
```diff
@@ -159,3 +159,13 @@ def DECIMAL128_to_decimal(self, int128_bytes: bytes, scale: int) -> decimal.Deci
         digits = [int(digit) for digit in str(int128) if digit != "-"]
         sign = int128 < 0
         return decimal.Decimal((sign, digits, -scale))
+
+    def DECFLOAT_to_decimal(self, exponent: int, significand: bytes) -> decimal.Decimal:
+        # significand is two's complement big endian. Convert to target endianess if needed.
+        if byteorder == "little":
+            significand = reversed(significand)
+        significand = int.from_bytes(significand, byteorder=byteorder, signed=True)
```
Comment (sfc-gh-pfus) [inline src/snowflake/connector/arrow_context.py:167]: Is it correct? You reverse byte order if needed, but pass byteorder without changing?

Comment (sfc-gh-tkissinger) [reply src/snowflake/connector/arrow_context.py:167]: @sfc-gh-pfus If I got it right, the byteorder indicates the endianess of the machine the driver is running on. The byte array always arrives in big endian and the idea is to convert it to the machines preferred endianess to enable machine-specific optimizations if they are implemented.

Comment (sfc-gh-pfus) [reply src/snowflake/connector/arrow_context.py:167]: Ah, I see. So, is Python's endianness OS-specific? Java, for instance, internally always uses big-endian.

Comment (sfc-gh-tkissinger) [reply src/snowflake/connector/arrow_context.py:167]: @sfc-gh-pfus I'm not a python expert 🙂 Just know that you need to pass in the array and tell the constructor the byte order of the bytes in that array. If you check the function above, e.g. for parsing our fixed-point number as Decimal (DECIMAL128_to_decimal), we pass the machine endianess in. Arrow uses the machine endianess too when writing the results, but unfortunately GS and driver run on different machines where endianess could differ. So, the function above will return wrong results when we have the driver on a BE machine and GS on a LE machine. Think there is some flag in the arrow metadata telling us about the endianess it was created with, but we ignore this as it looks. So, something we should fix. For the DECFLOAT, we know it is always BE. This means we could tell the constructor to that it is BE w/o reversing the array. However, sometime there is an optimized code path if endianess matches the machine endianess, and that's why we adjust this here. But I don't know whether the is such an optimization in Python

File: src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecFloatConverter.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  if (stringView.size_bytes > 16) {
+    std::string errorInfo = Logger::formatString(
+        "[Snowflake Exception] only precisions up to 38 supported. "
+        "Please update to a newer version of the connector.");
+    logger->error(__FILE__, __func__, __LINE__, errorInfo.c_str());
+    PyErr_SetString(PyExc_Exception, errorInfo.c_str());
+    return nullptr;
+  }
+  PyObject* significand =
+      PyBytes_FromStringAndSize(stringView.data, stringView.size_bytes);
```
Comment (sfc-gh-mmishchenko) [inline src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecFloatConverter.cpp:75]: Have you considered building a significand as a Py-integer right here? I.e. to `PyLong_FromNativeBytes` instead, it accepts source endianness.

File: src/snowflake/connector/nanoarrow_cpp/ArrowIterator/SnowflakeType.cpp [hunk 3]
```diff
@@ -17,6 +17,7 @@ std::unordered_map<std::string, SnowflakeType::Type>
         {"DOUBLE PRECISION", SnowflakeType::Type::REAL},
         {"DOUBLE", SnowflakeType::Type::REAL},
         {"FIXED", SnowflakeType::Type::FIXED},
+        {"DECFLOAT", SnowflakeType::Type::DECFLOAT},
```
Comment (sfc-gh-mmishchenko) [inline src/snowflake/connector/nanoarrow_cpp/ArrowIterator/SnowflakeType.cpp:20]: Looks this whole list has been following alphabetic order so far, so this one should be the next after `DATE`.

**Review by sfc-gh-mmishchenko — COMMENTED** [review]
Just to check - will it also be possible also to cover with test negative scenarios, i.e. a wrong DECFLOAT schema in an input or too long significand?

File: src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecFloatConverter.cpp (hunk 2 above)
Comment (sfc-gh-tkissinger) [reply src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecFloatConverter.cpp:75]: we wanted to get out of the C++ world as early as possible and I see no advantage using this function over the python version.

Comment (sfc-gh-mmishchenko) [reply src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecFloatConverter.cpp:78]: Won't it do 1 alloc/copy less, so in general perform faster?
